File: 1125. Smallest Sufficient Team/Solution.cpp
class Solution {
public:
    vector<int> smallestSufficientTeam(vector<string>& req_skills, vector<vector<string>>& people) {
        int n = req_skills.size(), m = people.size();
        unordered_map<string, int> skill;
        for (int i = 0; i < n; i++) skill[req_skills[i]] = i;
        vector<int> pmask(m, 0);
        unordered_map<int, vector<int>> dp;
        for (int i = 0; i < (1 << n); i++) {
            vector<int> d;
            dp[i] = d;
        }
        for (int i = 0; i < m; i++) {
            for (const auto& c : people[i]) {
                pmask[i] |= (1 << skill[c]);
            }
        }
        for (int i = 0; i < m; i++) {
            for (const auto& [k, v] : dp) {
                if (k != 0 && v.size() == 0) continue;
                int curr = k | pmask[i];
                if (curr == k) continue;
                if (dp[curr].size() == 0 || dp[curr].size() > v.size() + 1) {
                    vector<int> nteam;
                    for (const auto& t : v) nteam.push_back(t);
                    nteam.push_back(i);
                    dp[curr] = nteam;
                }
            }
        }
        return dp[(1 << n) - 1];
    }
};

File: 1125. Smallest Sufficient Team/Solution.java
class Solution {
    public int[] smallestSufficientTeam(String[] req_skills, List<List<String>> people) {
        int n = req_skills.length, m = people.size();
        Map<String, Integer> skill = new HashMap<>();
        for (int i = 0; i < n; i++) skill.put(req_skills[i], i);
        int[] pmask = new int[m];
        Map<Integer, List<Integer>> dp = new HashMap<>();
        for (int i = 0; i < (1 << n); i++) dp.put(i, new ArrayList<>());
        for (int i = 0; i < m; i++) {
            for (String c : people.get(i)) {
                pmask[i] |= (1 << skill.get(c));
            }
        }
        for (int i = 0; i < m; i++) {
            for (Map.Entry<Integer, List<Integer>> pair : dp.entrySet()) {
                if (pair.getKey() != 0 && pair.getValue().size() == 0) continue;
                int curr = pair.getKey() | pmask[i];
                if (curr == pair.getKey()) continue;
                if (dp.get(curr).size() == 0 || dp.get(curr).size() > pair.getValue().size() + 1) {
                    List<Integer> nteam = new ArrayList<>(pair.getValue());
                    nteam.add(i);
                    dp.put(curr, nteam);
                }
            }
        }
        int[] res = new int[dp.get((1 << n) - 1).size()];
        for (int i = 0; i < dp.get((1 << n) - 1).size(); i++) res[i] = dp.get((1 << n) - 1).get(i);
        return res;        
    }
}
